moped-editor/src/views/projects/projectView/ToolbarPlugin.js

```diff
@@ -0,0 +1,290 @@
+import { INSERT_ORDERED_LIST_COMMAND, INSERT_UNORDERED_LIST_COMMAND, REMOVE_LIST_COMMAND } from "@lexical/list";
+import { TOGGLE_LINK_COMMAND, $isLinkNode } from "@lexical/link";
+
+const richTextAction = {
+  bold: "bold",
+  italics: "italics",
+  underline: "underline",
+  number: "number",
+  bullet: "bullet",
+  link: "link",
+  divider: "divider",
+  undo: "undo",
+  redo: "redo",
+  clear: "clear"
+}
+
+const RICH_TEXT_OPTIONS = [
+  { id: richTextAction.bold, icon: <FormatBold />, label: "Bold" },
+  { id: richTextAction.italics, icon: <FormatItalic />, label: "Italics" },
+  { id: richTextAction.underline, icon: <FormatUnderlined />, label: "Underline" },
+  { id: richTextAction.divider },
+  {
+    id: richTextAction.number,
+    icon: <FormatListNumbered />,
+    label: "Numbered list",
+  },
+  {
+    id: richTextAction.bullet,
+    icon: <FormatListBulleted />,
+    label: "Bulleted list",
+  },
+  {
+    id: richTextAction.link,
+    icon: <Link />,
+    label: "Link",
+  },
+  { id: richTextAction.divider },
+  {
+    id: richTextAction.undo,
+    icon: <Undo />,
+    label: "Undo",
+  },
+  {
+    id: richTextAction.redo,
+    icon: <Redo />,
+    label: "Redo",
+  },
+  {
+    id: richTextAction.clear,
+    icon: <FormatClear />,
+    label: "Clear formatting",
+  },
+];
+
+const ToolbarPlugin = ({ noteAddSuccess, classes }) => {
+  const [editor] = useLexicalComposerContext();
+
+  const [disableMap, setDisableMap] = useState({
+    [richTextAction.undo]: true,
+    [richTextAction.redo]: true
+  });
+  const [selectionMap, setSelectionMap] = useState({});
+
+  // Iterate through selected node and parent nodes and returns any applicable list type
+  const checkListType = (selection, listType) => {
+    let hasListType = false;
+    const selectedNodes = selection.getNodes();
+    selectedNodes.forEach((node) => {
+      const nodeParents = node.getParents();
+      nodeParents.forEach((parent) => {
+        if (parent.__listType === listType) {
+          hasListType = true;
+        }
+      })
+    })
+    return hasListType;
+  };
+
+  // Checks for selected list formatting, removes if already applied or adds if not applied
+  const handleListUpdate = (command, listType) => {
+    const isListApplied = selectionMap[listType];
+    isListApplied ?
+      editor.dispatchCommand(REMOVE_LIST_COMMAND) :
+      editor.dispatchCommand(command);
+  };
+
+  const checkLink = (selection) => {
+    let isLink = false;
+    const selectedNodes = selection.getNodes();
+    selectedNodes.forEach((selectedNode) => {
+      const selectedNodeParent = selectedNode.getParent()
+      if ($isLinkNode(selectedNodeParent)) {
+        isLink = true
+      }
+    })
+    return isLink;
+  };
+
+  const handleLinkUpdate = () => {
+    const isLinkApplied = selectionMap["link"];
+    isLinkApplied ?
+      editor.dispatchCommand(TOGGLE_LINK_COMMAND, null)
+      :
+      navigator.clipboard.readText().then(val => {
```

Comment:
I stumbled a little when trying to create a link and unintentionally captured my clipboard to a link. I like this ternary to toggle between link and no link, and I'm wondering the toggle could be between:
1. If highlighted text isn't already a link, the text turns into a link (whether it is valid or invalid - just like production code which doesn't worry about validating what is entered as a link)
2. If highlighted text is already a link, the link is removed from the text

I think that the editor dispatches could happen in the context of `editor.update` so that you can access `selection.getTextContent()` and use that to set the `url` value. Overall, I'm worried about introducing a more complex interaction that involves the clipboard. 100% open to other opinions here in case I'm missing something!

Reply:
after chatting about this @mddilley and i decided to stick with this behavior for now and create a follow-up issue to add a modal popup for pasting and removing links to optimize usability of this feature.

Reply:
sounds good—i'm definitely not a fan of the behavior that tries to read from the clipboard.

Reply:
i just tested this updated behavior, and this looks great! Thanks Tilly! 🚀

Reply:
@tillyw I just accidentally saved a new Moped user's email to a link, and I think people will be caught off guard by this button interacting with their clipboard. I think that this should update to toggle between link/non-link when text is highlighted to prevent us from confusing users. It could even lead them to capture something sensitive that they don't into to into the app. I included a snippet of what I think it would take and a clip of the resulting behavior.

Could you update? Copying and pasting links like AMD often does works beautifully! ✨

```js
editor.update(() => {
   const isLinkApplied = selectionMap["link"];
   const selection = $getSelection();
   const textContent = selection.getTextContent();

   isLinkApplied
     ? editor.dispatchCommand(TOGGLE_LINK_COMMAND, null) // Remove link if already applied
     : editor.dispatchCommand(TOGGLE_LINK_COMMAND, textContent); // Make selected text into a link
 });
```

https://github.com/user-attachments/assets/1f54bbfa-ea35-4198-93f3-e68d74aa726f

Reply:
this works well! will go with this for now